[17.0][ADD] privacy_partner_to_be_forgotten-add_gdpr_module

halbtonjazz opened the pull request:
This PR adds a new module that implements GDPR compliance features for contact anonymization in Odoo.

Key features:
- Anonymizes PII in res.partner records (name, contact details, documents)
- Preserves company relationships while anonymizing individual contacts
- Handles linked res.users records
- Removes associated messages and attachments
- Adds traceability through log notes
- Supports multi-level contact hierarchies

The anonymization process:
- Replaces names with anonymized initials
- Creates unique anonymized emails with timestamps and user id
- Clears all personal contact information
- Archives the records
- Removes all chatter history
- Handles nested contact structures

Security:
- Requires specific user permissions
- Shows confirmation wizard
- Provides clear warnings about irreversible changes

Task: 4518

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced a module allowing anonymization of contact records to support GDPR compliance, including removal of personal data, messages, attachments, and anonymization of linked user accounts.
  - Added a user permission system to control access to anonymization functionality.
  - Provided a wizard interface in the Contacts app for selecting and anonymizing partners, with warnings and confirmation steps.
  - Added action in partner views to trigger anonymization wizard.

- **Documentation**
  - Added comprehensive documentation covering module description, configuration, usage instructions, context, and contributor information.
  - Included HTML and markdown files for detailed guidance and compliance context.

- **Tests**
  - Implemented automated tests to verify anonymization behavior, access control, and data integrity.

- **Chores**
  - Added module manifest and build configuration files for deployment and packaging.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

halbtonjazz committed:
[ADD] privacy_partner_to_be_forgotten: draft module

Add initial draft of the GDPR anonymization module to comply
with the "Right to be Forgotten" for individual contact records
(res.partner).

Includes base structure, wizard logic, access control, and UI
integration for manual anonymization via action menu.

Task: 4518

halbtonjazz committed:
[REF] privacy_partner_to_be_forgotten: refactor

Improve module implementation with demo data and code refinements:
- Add comprehensive demo data with test partners and users
- Refactor anonymization logic for better maintainability
- Update security groups with dedicated category
- Improve wizard UI and warning messages
- Fix mail.channel references to discuss.channel
- Add no_reset_password context for demo users
- Clean up docstrings and code formatting

Task: 4518

halbtonjazz committed:
[IMP] privacy_partner_to_be_forgotten: formatting & context

Improve XML formatting and context handling in the module:
- Add proper XML formatting
- Include active_test=False in partner anonymization context to handle
  archived records

These changes enhance code readability and ensure proper handling of both
active and archived partners during anonymization process.

Task: 4518

halbtonjazz committed:
[ADD] privacy_partner_to_be_forgotten: docs

Add initial documentation for the module including:
- README and subpages (CONTEXT, USAGE, CONFIGURE, etc.)
- Contributors list in readme
- HTML description for Apps Store display
- Updated manifest summary for better clarity
- Added copyright headers

Task: 4518

geomer198 commented on privacy_partner_to_be_forgotten/models/res_partner.py:
```diff
@@ -0,0 +1,134 @@
+        name_parts = (self.name or "").split()
+        initials = "".join(part[0].upper() for part in name_parts if part) or "X"
+
+        anonymized_name = f"{initials} Anonymized"
```
I think better use translation string  `_()` here.

halbtonjazz replied:
I've made that change, thank you

geomer198 commented on privacy_partner_to_be_forgotten/models/res_partner.py:
```diff
@@ -0,0 +1,134 @@
+            {
+                "name": anonymized_name,
+                "email": anonymized_email,
+                "phone": False,
```
Better use separate method for preparing updating values for partner.
For example:
```py
def _preapre_partner_anonymized_vals():
    ...
```

halbtonjazz replied:
I realized, did it

geomer198 commented on privacy_partner_to_be_forgotten/models/res_partner.py:
```diff
@@ -0,0 +1,134 @@
+        # Anonymize linked user accounts
+        self.user_ids.write(
+            {
+                "login": anonymized_email,
```
Here too

halbtonjazz replied:
done

geomer198 commented on privacy_partner_to_be_forgotten/models/res_partner.py:
```diff
@@ -0,0 +1,134 @@
+class ResPartner(models.Model):
+    _inherit = "res.partner"
+
+    def anonymize_partner_data(self):
```
This method is too complex. Please split method implementation  on the methods. This methods maybe will updated in different modules.

Methods `_anonymize_partner` `_anonymize_partner_message` `_anonymize_partner_attachment`

halbtonjazz replied:
I've accepted these changes, thanks

geomer198 commented on privacy_partner_to_be_forgotten/models/res_partner.py:
```diff
@@ -0,0 +1,134 @@
+        )
+
+        # Delete all messages related to this partner
+        message_domain = [
```
Create method for preparing domain for unlink messages.

halbtonjazz replied:
Done

geomer198 commented on privacy_partner_to_be_forgotten/models/res_partner.py:
```diff
@@ -0,0 +1,134 @@
+        messages = self.env["mail.message"].sudo().search(message_domain)
+
+        # Delete attachments
+        attachment_domain = [
```
Here too

halbtonjazz replied:
Done

geomer198 commented on privacy_partner_to_be_forgotten/models/res_partner.py:
```diff
@@ -0,0 +1,134 @@
+        messages = self.env["mail.message"].sudo().search(message_domain)
+
+        # Delete attachments
+        attachment_domain = [
```
Here too

halbtonjazz replied:
ok, done

geomer198 commented on privacy_partner_to_be_forgotten/wizards/partner_anonymize_wizard.py:
```diff
@@ -0,0 +1,63 @@
+
+    partner_ids = fields.Many2many(
+        "res.partner",
+        string="Partners",
```
This string is duplicated prepared `string` value by field name. Please remove this line.

halbtonjazz replied:
I got it

geomer198 commented on privacy_partner_to_be_forgotten/wizards/partner_anonymize_wizard.py:
```diff
@@ -0,0 +1,63 @@
+            raise AccessError(_("You don't have permission to anonymize partners."))
+
+        # Check if partner is a company
+        company_partners = self.partner_ids.filtered(lambda p: p.is_company)
```
Please use filtered with string parameter.
```py
self.partner_ids.filtered("is_company")
```

halbtonjazz replied:
Fixed, thanks

geomer198 commented on privacy_partner_to_be_forgotten/models/res_partner.py:
```diff
@@ -0,0 +1,134 @@
+        self.ensure_one()
+
+        # Generate anonymized values
+        now = datetime.datetime.now()
```
You can use `fields.Datetime.now()` here,

halbtonjazz replied:
Ok, thanks, I read up on the difference

geomer198 commented on privacy_partner_to_be_forgotten/models/res_partner.py:
```diff
@@ -0,0 +1,134 @@
+
+        # Generate anonymized values
+        now = datetime.datetime.now()
+        date_stamp = now.strftime("%d.%m.%y")
```
You can use `to_string` from `fields.Datetime`

halbtonjazz replied:
fields.Datetime.to_string() returns the date in the format YYYYY-MM-DD HH:MM:SS, and we need a specific format DD.MM.YY to generate an anonymized email address.
In the current code we use strftime(“%d.%m.%y”) exactly to get the specific date format

geomer198 commented on privacy_partner_to_be_forgotten/models/res_partner.py:
```diff
@@ -0,0 +1,134 @@
+        attachments = self.env["ir.attachment"].sudo().search(attachment_domain)
+
+        # Delete messages and attachments
+        messages.unlink()
```
I think better add logger for check which records was removed.

halbtonjazz replied:
Sure, no problem, done. Thanks for idea

geomer198 commented on privacy_partner_to_be_forgotten/models/res_partner.py:
```diff
@@ -0,0 +1,134 @@
+
+        # Add a log note about anonymization
+        self.message_post(
+            body=_("This contact has been anonymized on %(date)s by %(user)s")
```
For 17 version better use translation string like this:
```py
_("... %(date)s ...", date=now.strftime("%Y-%m-%d %H:%M:%S"))
```

halbtonjazz replied:
Done

geomer198 commented on privacy_partner_to_be_forgotten/HISTORY.md:
```diff
@@ -0,0 +1 @@
+
```
This file must be located in `readme` directory

geomer198 commented on privacy_partner_to_be_forgotten/__manifest__.py:
```diff
@@ -0,0 +1,23 @@
+        "wizards/partner_anonymize_wizard_views.xml",
+    ],
+    "demo": ["demo/demo_partner_anonymize_data.xml"],
+    "installable": True,
```
Please remove `installable` and `application` because this keys have default value.

halbtonjazz replied:
I get why, thanks. Done

geomer198 reviewed (commented):
Please read the code commentary and add tests.

halbtonjazz committed:
[REF] privacy_partner_to_be_forgotten: refactor + tests

Refactor anonymization logic to improve readability,
testability, and compliance with Odoo conventions.

Changes include:
- Extracted helper methods for email, messages, attachments
- Replaced datetime with fields.Datetime for consistency
- Improved logging and translation usage
- Rewrote _anonymize_partner_data into modular structure

Added extensive test coverage:
- Parent/child anonymization scenarios
- Company restriction handling
- Access rights validation
- Attachments/messages cleanup
- Multi-partner anonymization

Minor improvements:
- Removed unused demo import
- Removed optional manifest keys
- Added changelog section to docs

Task: 4518

geomer198 commented on privacy_partner_to_be_forgotten/demo/demo_partner_anonymize_data.xml:
```diff
@@ -0,0 +1,142 @@
+    Copyright (C) 2025 Cetmix OÜ
+    License LGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+-->
+<odoo>
```
Better will be add `noupdate="1"`.

halbtonjazz replied:
I suggest not adding noupdate="1" because:
1. These are demo data (not initialization data) specifically for tests
2. We want these records to be recreated on each module update to ensure clean test state
3. Tests modify these records (setting parent_id, anonymizing partners), so we need fresh data on each test run

geomer198 replied:
@halbtonjazz  We are not use demo data in the tests, bad practice.

halbtonjazz replied:
I fixed that comment in a new commit for the new PR, but haven't pushed it yet

geomer198 commented on privacy_partner_to_be_forgotten/models/res_partner.py:
```diff
@@ -0,0 +1,173 @@
+class ResPartner(models.Model):
+    _inherit = "res.partner"
+
+    def _get_partner_initials(self):
```
Better make this method like `@property` `get_partner_initials` or create separate field with stored and compute method.

halbtonjazz replied:
I suggest keeping the current implementation as `_get_partner_initials()` method because:
1. It's a private helper method used only once during the anonymization process
2. Converting it to `@property` or computed field would be excessive since:
   - We don't need to store/compute initials constantly
   - Initials are not used in views or other parts of the code
   - It would add unnecessary database overhead
3. Current implementation follows YAGNI principle and clearly indicates its helper function with `_` prefix

geomer198 commented on privacy_partner_to_be_forgotten/models/res_partner.py:
```diff
@@ -0,0 +1,173 @@
+        date_stamp = fields.Date.today().strftime("%d.%m.%y")
+        return f"{initials.lower()}_{date_stamp}_{self.id}@anonymized.oca"
+
+    def _prepare_partner_anonymized_vals(self, anonymized_name, anonymized_email):
```
Better make arguments like a `**kwards` and add to result:
```py
return {
    **kwards,
    ...
}
```

halbtonjazz replied:
Thank you for the suggestion.
In this case, I’d prefer to keep the explicit signature for better clarity.
Since the method only ever uses name and email, passing them directly avoids possible runtime issues from missing keys and keeps the interface more readable.
Using **kwargs would:
   - Make the code less explicit
   - Make it harder to track what values are actually being passed
   - Complicate static code analysis
Current implementation follows "Explicit is better than implicit" principle

geomer198 commented on privacy_partner_to_be_forgotten/models/res_partner.py:
```diff
@@ -0,0 +1,173 @@
+            "active": False,
+        }
+
+    def _prepare_user_anonymized_vals(self, anonymized_email):
```
Here too.

halbtonjazz replied:
Same here

geomer198 commented on privacy_partner_to_be_forgotten/models/res_partner.py:
```diff
@@ -0,0 +1,173 @@
+        """Anonymize related user accounts."""
+        if self.user_ids:
+            self.user_ids.write(self._prepare_user_anonymized_vals(anonymized_email))
+            self.user_ids = self.env["res.users"].browse(self.user_ids.ids)
```
For what do you made this?

halbtonjazz replied:
This code serves two purposes:
1. Main purpose: It anonymizes all user accounts (res.users) linked to the partner as part of GDPR compliance. When a partner requests to be "forgotten", we must also deactivate and anonymize all their system accounts.
2. Technical aspect: The last line (self.user_ids = self.env["res.users"].browse(self.user_ids.ids)) ensures cache consistency after writing critical user fields (login, email). I added it after observing cache inconsistencies during testing.
I can remove the cache refresh line if you think it's redundant in Odoo 17, but included it as a safeguard.

halbtonjazz replied:
I decided to remove self.user_ids = self.env[“res.users”].browse(self.user_ids.ids) thinking this code was redundant. I fixed that comment in a new commit for the new PR, but haven't pushed it yet.

geomer198 commented on privacy_partner_to_be_forgotten/models/res_partner.py:
```diff
@@ -0,0 +1,173 @@
+        return f"{initials.lower()}_{date_stamp}_{self.id}@anonymized.oca"
+
+    def _prepare_partner_anonymized_vals(self, anonymized_name, anonymized_email):
+        """Prepare values for anonymizing partner data."""
```
Please describe arguments in docstring.

halbtonjazz replied:
Done

geomer198 commented on privacy_partner_to_be_forgotten/models/res_partner.py:
```diff
@@ -0,0 +1,173 @@
+        return initials or "X"
+
+    def _generate_anonymized_email(self, initials):
+        """Generate anonymized email address."""
```
Please describe arguments in docstring.

halbtonjazz replied:
Done

geomer198 commented on privacy_partner_to_be_forgotten/models/res_partner.py:
```diff
@@ -0,0 +1,173 @@
+        }
+
+    def _prepare_user_anonymized_vals(self, anonymized_email):
+        """Prepare values for anonymizing related user accounts."""
```
Here too.

halbtonjazz replied:
Done

geomer198 commented on privacy_partner_to_be_forgotten/models/res_partner.py:
```diff
@@ -0,0 +1,173 @@
+        }
+
+    def _anonymize_user(self, anonymized_email):
+        """Anonymize related user accounts."""
```
Here too and next docstrings

halbtonjazz replied:
Done

geomer198 commented on privacy_partner_to_be_forgotten/models/res_partner.py:
```diff
@@ -0,0 +1,173 @@
+        """Prepare domain for searching attachments to be deleted during
+        anonymization.
+        """
+        return [
```
Please add `self.ensure_one()` or use `.ids` in domain

halbtonjazz replied:
Done

geomer198 commented on privacy_partner_to_be_forgotten/models/res_partner.py:
```diff
@@ -0,0 +1,173 @@
+        self.ensure_one()
+        now = fields.Datetime.now()
+
+        initials = self._get_partner_initials()
+        anonymized_name = _("%s Anonymized") % initials
```
@halbtonjazz Please check this comment and next's comments from coderabbitai.

halbtonjazz replied:
taken, done

geomer198 commented on privacy_partner_to_be_forgotten/wizards/partner_anonymize_wizard.py:
```diff
@@ -0,0 +1,73 @@
+    @api.model
+    def default_get(self, fields_list):
+        res = super().default_get(fields_list)
+        if "partner_ids" in fields_list:
```
You can make one consition:
```py
active_ids = self._context.get("active_ids")
if "partner_ids" in field_list and active_ids:
    ...
```

halbtonjazz replied:
Updated to a single condition for readability. Thanks!

geomer198 commented:
@halbtonjazz Please respond to the bot on its suggestions for changes.

halbtonjazz commented on privacy_partner_to_be_forgotten/wizards/partner_anonymize_wizard.py:
```diff
@@ -0,0 +1,73 @@
+                % anonymized_names
+            )
+
+    def action_confirm(self):
+        """Confirm and process partner anonymization for multiple partners"""
+        self.ensure_one()
+        self._validate_partners_for_anonymization()
+
+        for partner in self.partner_ids:
+            partner.anonymize_partner_data()
+
+        return {
+            "type": "ir.actions.client",
+            "tag": "display_notification",
+            "params": {
+                "title": _("Anonymization Result"),
+                "message": _("Partner(s) have been anonymized successfully."),
+                "sticky": False,
+                "type": "success",
+                "next": {"type": "ir.actions.act_window_close"},
+            },
+        }
```
Good suggestion, I will apply it. Thanks

halbtonjazz commented on privacy_partner_to_be_forgotten/static/description/index.html:
```diff
@@ -0,0 +1,523 @@
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+!! source digest: sha256:9961888c350357b97ad3faa7c8d0b96f0603f938f56567845cbcaf7f01a7a4a0
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! -->
+<p><a class="reference external image-reference" href="https://odoo-community.org/page/development-status"><img alt="Beta" src="https://img.shields.io/badge/maturity-Beta-yellow.png" /></a> <a class="reference external image-reference" href="http://www.gnu.org/licenses/lgpl-3.0-standalone.html"><img alt="License: LGPL-3" src="https://img.shields.io/badge/licence-LGPL--3-blue.png" /></a> <a class= ...
```
automatically generated by the oca-gen-addon-readme tool

halbtonjazz committed:
[REF] privacy_partner_to_be_forgotten: docstrings + fixes

Improve code documentation and formatting across multiple files to enhance
readability and maintainability:

- Add detailed docstrings to core anonymization methods
- Use named placeholders in translations for better clarity
- Add ensure_one() safety check to attachment domain method
- Improve code formatting and line length compliance
- Simplify default_get logic in wizard
- Enhance anonymization success message with count

Technical improvements:
- Use rf-strings for email regex patterns
- Standardize create() and search() formatting

Task: 4518

halbtonjazz commented on privacy_partner_to_be_forgotten/wizards/partner_anonymize_wizard.py:
```diff
@@ -0,0 +1,73 @@
+# Copyright (C) 2025 Cetmix OÜ
+# License LGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
```
Good point, thanks

halbtonjazz commented on privacy_partner_to_be_forgotten/tests/test_partner_anonymize.py:
```diff
@@ -0,0 +1,415 @@
+# Copyright (C) 2025 Cetmix OÜ
+# License LGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
```
Good point, thanks

halbtonjazz commented on privacy_partner_to_be_forgotten/models/res_partner.py:
```diff
@@ -0,0 +1,206 @@
+        """
+        if self.user_ids:
+            self.user_ids.write(self._prepare_user_anonymized_vals(anonymized_email))
+            self.user_ids = self.env["res.users"].browse(self.user_ids.ids)
```
I am discussing this point with the reviewer, it may be an redundant line of code

halbtonjazz commented on privacy_partner_to_be_forgotten/models/res_partner.py:
```diff
@@ -0,0 +1,206 @@
+# Copyright (C) 2025 Cetmix OÜ
+# License LGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
```
Done

halbtonjazz committed:
[FIX] privacy_partner_to_be_forgotten: fix license URLs

Update all license header URLs to correctly point to LGPL-3.0 instead of AGPL
across the module. This ensures consistency between the declared license
(LGPL-3.0) and its reference URL.

Files updated:
- Python files (__init__.py, __manifest__.py, models, tests, wizards)
- XML files (demo data, security, views)

This change improves legal accuracy and maintains proper license
documentation throughout the codebase.

Task: 4518

halbtonjazz commented on privacy_partner_to_be_forgotten/security/security.xml:
```diff
@@ -0,0 +1,23 @@
+    </record>
+
+    <!-- Define the group for Anonymize Contacts -->
+    <record id="group_partner_anonymize" model="res.groups">
+        <field name="name">Administrator</field>
+        <field name="category_id" ref="module_category_anonymize_contacts" />
+        <field name="users" eval="[(4, ref('base.user_admin'))]" />
+        <field
+            name="comment"
+        >Users in this group can anonymize contact information for GDPR compliance.</field>
+    </record>
```
I'll discuss it with the manager

halbtonjazz commented on privacy_partner_to_be_forgotten/wizards/partner_anonymize_wizard.py:
```diff
@@ -0,0 +1,73 @@
+            res["partner_ids"] = [(6, 0, all_partners.ids)]
+        return res
+
+    def _validate_partners_for_anonymization(self):
+        # Check access rights
+        if not self.env.user.has_group(
+            "privacy_partner_to_be_forgotten.group_partner_anonymize"
+        ):
+            raise AccessError(_("You don't have permission to anonymize partners."))
+
+        # Check if partner is a company
+        company_partners = self.partner_ids.filtered("is_company")
+        if company_partners:
+            company_names = ", ".join(company_partners.mapped("name"))
+            raise UserError(
+                _("Cannot anonymize the following company records: %s") % company_names
+            )
+
```
I think that's a good suggestion, I'll discuss it with the reviewer

halbtonjazz commented on privacy_partner_to_be_forgotten/wizards/partner_anonymize_wizard.py:
```diff
@@ -0,0 +1,73 @@
+        readonly=True,
+    )
+
+    @api.model
+    def default_get(self, fields_list):
+        res = super().default_get(fields_list)
+        active_ids = self._context.get("active_ids")
+        if "partner_ids" in fields_list and active_ids:
+            domain = [("id", "child_of", active_ids)]
+            all_partners = self.env["res.partner"].search(domain)
+            res["partner_ids"] = [(6, 0, all_partners.ids)]
+        return res
```
Thanks for the suggestion. I prefer to keep the current implementation as it was not specified in the terms of reference. Perhaps after functional testing we will come to a similar solution.

halbtonjazz commented on privacy_partner_to_be_forgotten/models/res_partner.py:
```diff
@@ -0,0 +1,206 @@
+            "signature": False,
+        }
+
+    def _anonymize_user(self, anonymized_email):
+        """Anonymize related user accounts.
+
+        Args:
+            anonymized_email (str): Generated anonymous email to be set
+                for user accounts
+        """
+        if self.user_ids:
+            self.user_ids.write(self._prepare_user_anonymized_vals(anonymized_email))
+            self.user_ids = self.env["res.users"].browse(self.user_ids.ids)
+
```
discussing with reviewer self.user_ids = self.env[“res.users”].browse(...)
I think yes, we'll probably remove it

halbtonjazz commented on privacy_partner_to_be_forgotten/wizards/partner_anonymize_wizard.py:
```diff
@@ -0,0 +1,73 @@
+                % anonymized_names
+            )
+
+    def action_confirm(self):
+        """Confirm and process partner anonymization for multiple partners"""
+        self.ensure_one()
+        self._validate_partners_for_anonymization()
+
+        for partner in self.partner_ids:
+            partner.anonymize_partner_data()
+
```
thanks for the suggestion, I prefer to keep the current implementation

halbtonjazz commented on privacy_partner_to_be_forgotten/demo/demo_partner_anonymize_data.xml:
```diff
@@ -0,0 +1,142 @@
+        <field name="function">Student</field>
+    </record>
+
+    <!-- Test Users -->
+    <record
+        id="user_company_sodasopa"
+        model="res.users"
+        context="{'no_reset_password': True}"
+    >
+        <field name="partner_id" ref="partner_company_sodasopa" />
+        <field name="login">sodasopa</field>
+        <field name="password">sodasopa</field>
+        <field name="name">Sodasopa</field>
+        <field name="email">info@sodasopa.com</field>
+        <field name="groups_id" eval="[(4, ref('base.group_user'))]" />
+        <field name="action_id" ref="mail.action_discuss" />
+    </record>
```
I'll discuss it with the reviewer

halbtonjazz commented on privacy_partner_to_be_forgotten/models/res_partner.py:
```diff
@@ -0,0 +1,206 @@
+            "active": False,
+        }
+
+    def _prepare_user_anonymized_vals(self, anonymized_email):
+        """Prepare values for anonymizing related user accounts.
+
+        Args:
+            anonymized_email (str): Generated anonymous email
+                (format: "xx_dd.mm.yy_id@anonymized.oca")
+
+        Returns:
+            dict: Dictionary containing user fields to update
+                (login, email, active status, signature)
+        """
+        return {
+            "login": anonymized_email,
+            "email": anonymized_email,
+            "active": False,
+            "signature": False,
+        }
+
```
I'll check it out.

halbtonjazz commented on privacy_partner_to_be_forgotten/models/res_partner.py:
```diff
@@ -0,0 +1,206 @@
+            str: Anonymized email address in format
+                "initials_dd.mm.yy_id@anonymized.oca"
+        """
+        date_stamp = fields.Date.today().strftime("%d.%m.%y")
+        return f"{initials.lower()}_{date_stamp}_{self.id}@anonymized.oca"
+
```
Repeated anonymization is impossible because it's archived during the first anonymization. I have already discussed this point with the manager. It was decided to leave this solution

halbtonjazz commented:
Closing this PR to reopen it from the original repository as requested.
